libraries/chain/include/eosio/chain/hotstuff.hpp

```diff
@@ -0,0 +1,86 @@
+#pragma once
+#include <eosio/chain/block_header.hpp>
+#include <fc/bitutil.hpp>
+#include <fc/crypto/bls_private_key.hpp>
+#include <fc/crypto/bls_public_key.hpp>
+#include <fc/crypto/bls_signature.hpp>
+#include <fc/crypto/bls_utils.hpp>
+
+namespace eosio::chain {
+
+   const block_id_type NULL_BLOCK_ID = block_id_type("00");
+   const fc::sha256 NULL_PROPOSAL_ID = fc::sha256("00");
+
+   inline uint64_t compute_height(uint32_t block_height, uint32_t phase_counter) {
+      return (uint64_t{block_height} << 32) | phase_counter;
+   }
+
+   struct extended_schedule {
+      producer_authority_schedule                          producer_schedule;
+      std::map<name, fc::crypto::blslib::bls_public_key>   bls_pub_keys;
+   };
+
+   struct quorum_certificate {
+      fc::sha256                          proposal_id = NULL_PROPOSAL_ID;
+      fc::unsigned_int                    active_finalizers = 0; //bitset encoding, following canonical order
+      fc::crypto::blslib::bls_signature   active_agg_sig;
+      bool                                quorum_met = false;
+   };
+
+   struct hs_vote_message {
+      fc::sha256                          proposal_id = NULL_PROPOSAL_ID; //vote on proposal
+      name                                finalizer;
+      fc::crypto::blslib::bls_signature   sig;
+   };
+
+   struct hs_proposal_message {
+      fc::sha256                          proposal_id = NULL_PROPOSAL_ID; //vote on proposal
+      block_id_type                       block_id = NULL_BLOCK_ID;
+      fc::sha256                          parent_id = NULL_PROPOSAL_ID; //new proposal
+      fc::sha256                          final_on_qc = NULL_PROPOSAL_ID;
+      quorum_certificate                  justify; //justification
```

Comment:
All these ` = NULL_PROPOSAL_ID` are unneeded since the default construction of sha256 already initializes to all zeroes.